### DISCUSSION TITLE
Building a project without a database connection

### DISCUSSION THREAD

**Discussion by wesdimiceli** [payloadcms/payload#9028] [Q&A]
I'm in the process of taking our existing production v2 implementation and working through the migration guide for v3.  One issue that I can't seem to resolve is getting a NextJS production build without needing to set the secret or Mongo connection details in the config.  Right now I'm basically looking at a stock v3 template and trying to run `pnpm build` gives me:

```
> cross-env NODE_OPTIONS=--no-deprecation NODE_ENV=production next build

  ▲ Next.js 15.0.0-canary.173
  - Experiments (use with caution):
    · turbo

   Creating an optimized production build ...
 ✓ Compiled successfully
 ✓ Linting and checking validity of types
   Collecting page data  ...Error: Error: missing secret key. A secret key is needed to secure Payload.

...

> Build error occurred
Error: Failed to collect page data for /admin/[[...segments]]
```

Even if I set a dummy secret, it will still fail on attempt to connect to Mongo, depending on what connection string is set in the local environment files.  Previously, we had the `disableDBConnect` option but it doesn't seem to be supported by the v3 config.

Feels to me like universal best practice is to avoid production secrets in the repo and/or sitting around in developers' .env files, but the example READMEs suggest something along the lines of "copy the .env.example file and fill in the blanks".  Which is fine for dev tasks but in order to get a usable docker image I need an environment-agnostic build that I can push to ECR, and putting secrets or database credentials in that image would be a pretty big security violation.

I'm not clear on what the database connection is used for during the build process, but I can imagine there could be a lot of instances where a DB connection is not practical -- e.g. DB lives in a private subnet, or some CI/CD configurations where a local database isn't feasible.  Is there any way to achieve what I'm looking to do here?  It would be great if there were some official documentation surrounding best practices w.r.t. using an encrypted secrets store like AWS SecretsManager or SSM.  As it stands, the current Dockerfile examples don't seem to cover how to handle production secrets.

**Comment by andershermansen** [comment]
I do not think payload itself and the admin panel requires the db connection.

But if you put any pages which uses payload to fetch data (like probably the templates does). Then it will need to contact payload to build those pages.

Try to remove those pages and see if it will build without then.

**Reply by wesdimiceli** [reply]
Thanks.  I don't think this is the case in my situation.  The admin's root layout initializes Payload which always tries to establish a DB connection.  When the connection fails, the build errors out.  That means some DB needs to be available even if it's not used in the build process.

**Comment by rob-at-cortex** [comment]
did this get resolved? we have the same issue

**Reply by wesdimiceli** [reply]
For me, I tracked it down to a component in the import map that was initializing payload.  I'd check there to see if anything might be calling into payload in your codebase.  Personally, I find these aspects of nextjs to be unintuitive so I guess it takes a little trial-and-error.

**Comment by spaaacetoast** [comment]
I also ran into this issue, specifically I was calling a server action from a server side component. what I didn't forsee was that it would call on the global instance of payload during the prerendering... Causing the build to fail :(

Thanks to you all I figured it out :)

**Reply by FinnIckler** [reply]
I am running into the same issue calling it from a server side component like
```
"use server";
import { getPayload } from "payload";
(...)
export default async function Navbar() {
  const payload = await getPayload({ config });
  const navbar = await payload.findGlobal({ slug: "nav" });
  (...)
 }
```
If that is not allowed, what did you do to fix this?

Edit: I am using `export const dynamic = "force-dynamic";` for now on all pages that need payload, but it feels a bit of a drastic solution

**Comment by SychMusti** [comment]
@wesdimiceli I am facing the same issue..Can you guide as to how to resolve the error?

**Reply by spaaacetoast** [reply]
For me the solution was to make sure payload is only called during runtime, seeing @wesdimiceli comment made me realize the culprit.. which was that I was using getPayload in a server action which would be called during the prerendering stage of nextjs.  In this stage there is no database so this would fail my build. 

I'm guessing @wesdimiceli was using a global getPayload in a module imported by the ImportMap making their build fail.

Perhaps you are running into the same issue of unknowingly accesing payload during the build process. Hope this helps a bit :)

**Reply by SychMusti** [reply]
@spaaacetoast Thanks for the insight man!

In the RootLayout, I’m thinking of conditionally loading the importMap and serverFunction based on NODE_ENV. During the build process (NODE_ENV=production), these would be set to null to avoid initializing Payload or making database connections.


```
const importMap = process.env.NODE_ENV === 'production' ? null : require('./admin/importMap').importMap;
const handleServerFunctions = process.env.NODE_ENV === 'production' ? null : require('@payloadcms/next/layouts').handleServerFunctions;

const Layout = ({ children }) => {
  const serverFunction = handleServerFunctions
    ? async function (args) {
        'use server';
        return handleServerFunctions({
          ...args,
          config,
          importMap,
        });
      }
    : null;

  return (
    <RootLayout
      config={config}
      importMap={importMap}
      serverFunction={serverFunction}
    >
      {children}
    </RootLayout>
  );
};
```

Am I on the right track?

**Reply by wesdimiceli** [reply]
So in my case, I was prototyping some stuff with server actions and had `getPayload` called in the module scope.  What tripped me was that the component in question wasn't being used -- but it was in the import map so I guess that was enough for it to be picked up during the build.

My understanding is that the import map is mainly used for dynamic server-side rendering of components, so it's doubly confusing that it would affect anything having to do with pre-rendering.  That said, my experience with nextjs is pretty limited so I don't have a lot of background to draw on.

---

> I’m thinking of conditionally loading the importMap and serverFunction based on NODE_ENV.

One issue to consider is your app is probably running with `NODE_ENV=production` in any kind of prod context, and I'm assuming the import map and server functions are critical components at runtime.  So if you conditionally load things that way, it might not work when you actually deploy the app.

What I would do is inspect the code in the import map etc and ensure you don't have anything that's dependent on actual data... rendering a list of models for example.

**Comment by LucienLeMagicien** [comment]
For the record, I had this issue while building my app in CI because I `throw` if some env vars (for example, the DB connection string) are not set - which is the case during build time in CI.

I worked around this by returning dummy strings **if** we're in NextJS's build mode. Helpfully NextJS sets an env var `NEXT_PHASE` just for that, as noted here: https://github.com/vercel/next.js/discussions/48736#discussioncomment-5704784, https://github.com/vercel/next.js/discussions/66832 (and somewhat https://nextjs.org/docs/app/api-reference/config/next-config-js#phase)

This is my helper:
```ts
import { PHASE_PRODUCTION_BUILD } from "next/constants";

export const getEnvOrThrow = (key: string) => {
  const isBuildtime = process.env.NEXT_PHASE === PHASE_PRODUCTION_BUILD;

  const value = process.env[key];
  if (!value) {
    if (isBuildtime) {
      // Return dummy data during build. Otherwise we'd throw in the "Collecting page data ..." step
      console.warn(`WARN: Returning dummy env value for "${key}"`);
      return "dummy string";
    } else {
      throw new Error(`Environment variable ${key} is not set`);
    }
  }
  return value;
};
```

It feels somewhat dirty, but the build completes now. :sweat_smile: It's fine as long as no code actually "runs" (i.e. something actually tries to hit the database) during build.
